[nss] Add new ports nspr and nss

plq opened the pull request:
**Describe the pull request**

- #### What does your PR fix?  

    Fixes #7171 #7172  

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  

    * I found it compiles on x64-windows, x86-windows using latest msvc
    * Added "=skip" entries for other platforms to ci.baseline.txt

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  

    I did not make separate pull requests for NSPR and NSS as these are quite tightly coupled. Otherwise, It looks fine to my eyes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  

    ~~I am still working on this PR~~ 
    yes

dg0yt commented:
Better don't touch `vcpkg_build_make` unless needed, or in a separate PR: Each push will trigger a CI rebuild of all ports that depend on this function, taking many hours.

JackBoosY commented:
Please get failure logs [here](https://dev.azure.com/vcpkg/public/_build/results?buildId=62710&view=artifacts&pathAsName=false&type=publishedArtifacts).
If you need some help, please ping me.

plq commented:
> Better don't touch `vcpkg_build_make` unless needed, or in a separate PR: Each push will trigger a CI rebuild of all ports that depend on this function, taking many hours.

done: #21296

plq commented:
Hey, thanks for your close interest.

> Please get failure logs [here](https://dev.azure.com/vcpkg/public/_build/results?buildId=62710&view=artifacts&pathAsName=false&type=publishedArtifacts). If you need some help, please ping me.

I don't have the resources to deal with all these platforms. The x64-windows dlls seem to work fine so far. I made sure the 32bit target builds but haven't tested the binaries. They should work, though. I can't comment about the others.

As for the logs, except for the spurious 502 from V8 source repo, the only issue I see is vcpkg_find_acquire forgetting about NSINSTALL. I don't see why it wouldn't work, perhaps you are limiting what CI runs from untrusted sources?

plq commented:
For completeness' sake, here's the stdout from x64-windows

```
-- Downloading https://releases.mozilla.org/pub/nspr/releases/v4.32/src/nspr-4.32.tar.gz -> nspr-4.32.tar.gz...
FATALunknown tool NSINSTALL -- unable to acquire.
CMake Error at scripts/cmake/vcpkg_find_acquire_program.cmake:493 (find_program):
  find_program called with incorrect number of arguments
Call Stack (most recent call first):
  scripts/cmake/vcpkg_find_acquire_program.cmake:534 (do_find)
  ports/nspr/portfile.cmake:14 (vcpkg_find_acquire_program)
  scripts/ports.cmake:142 (include)
```

JackBoosY commented on scripts/cmake/vcpkg_find_acquire_program.cmake:
```diff
@@ -425,6 +426,13 @@ function(vcpkg_find_acquire_program VAR)
+    set(PROGNAME nsinstall)
+    set(MOZBUILD_VERSION 3.3)
+    set(PATHS ${DOWNLOADS}/tools/nsinstall/bin)
+    set(URL "https://ftp.mozilla.org/pub/mozilla/libraries/win32/MozillaBuildSetup-${MOZBUILD_VERSION}.exe")
```
We can only use the compressed package, not installation file.

plq replied:
1. They don't publish anything else
2. I don't run it, I rename it to 7z.exe and unpack it.
3. There are other packages that do the same in that file.

JackBoosY replied:
I will take a look.

dg0yt replied:
Do we really want to add another program to `vcpkg_find_acquire_program.cmake`?
It is expensive, every change triggers rebuilds of all ports which depend on this function.
And I don't think `nsinstall` will become widely used.

plq commented:
I'll try to make your job easier by providing some questions to discuss:

1. If you don't want mozbuild in vcpkg_find_acquire_program (vfap), I can move it inside the ebuild. NSS needs GYP to build as well but I didn't put it inside vfap as it's a dead project and don't think it'll be useful elsewhere. mozbuild on the other hand is chromium's depot_tools and could come in handy.
    
    your choice, though.

2. I do weird things with headers: NSS has two sets of headers: public and private. public headers are put in ``include/nss/nss`` and private ones are put in ``include/nss/private/nss``. Users wouldn't notice anything with an accompanying FindNSS.cmake that deals with these quirks but I can't say I'm entirely happy with the current state of things. 

    so with all this said, what's vcpkg's policy regarding private headers?
    1. ignore them
    2. put them together with public headers
    3. put them in a subfolder inside public headers
    4. keep the status quo and offer them as a separate component

    thoughts?

3. The build also produces some .exe files. I put them inside bin but the QA bits complained so I removed them. Currently they are being ignored. What do you want me to do with them?

That's all I can think of. This one is may be second only to openssl in terms of weirdness of its build system :) Why do crypto libs are like this? They are the flakiest pieces of software we need to deal with, so need to be as easy to build as possible. 

:man_shrugging:

plq committed:
[nspr] Add new port 4.33

plq committed:
[nspr] Add new port 4.33

plq committed:
[nss] Add new port 3.73.1

plq committed:
[nss] Add new port 3.73.1

plq committed:
[nss] Add new port 3.73.1

plq commented:
the ebuilds were rebased on latest master, with mozbuild removed from vfap

JackBoosY commented:
Sorry for late, I will handle this tomorrow.

plq commented:
Here it says the latest stable release is 3.66: https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/NSS_Releases

There seem to be newer releases (newest is currently 3.72) and actually seems legitimate but couldn't be sure.

JackBoosY commented:
> I removed mozbuild from vcpkg_find_acquire_program (vfap)

Agreed, if the build system is too complex, we should add a new port instead of that.


> I do weird things with headers: NSS has two sets of headers: public and private. public headers are put in include/nss/nss and private ones are put in include/nss/private/nss. Users wouldn't notice anything with an accompanying FindNSS.cmake that deals with these quirks but I can't say I'm entirely happy with the current state of things.
\
so with all this said, what's vcpkg's policy regarding private headers?

If there is a call relationship between public header files and private header files, we still need to install them in the subfolders of _include_, because users actually need to use them.
If not, we should follow the upstream rules to choose whether to install them.


> The build also produces some .exe files. I put them inside bin but the QA bits complained so I removed them. Currently they are being ignored. What do you want me to do with them?

For those executables, they serve two purposes:
1. Generate to execute test cases, for them we should not install/generate or delete directly after installation
2. To generate tools, we should record their names and install them in _tools/${PORT}_ using `vcpkg_copy_tools`. See [documentation](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/vcpkg_copy_tools.md).


> It seems that the NSS build system also signs the produced binaries. It didn't work and I disabled it. I'm not sure it's supposed to work on random people's computers. I don't really care as we sign our installers anyway.

Yeah, some binaries, such as boost's build tool b2, also have this problem: the official binary is signed, but the code generated locally is not signed, which may cause the anti-virus report to be abnormal. The signature requires an official certificate, and there is nothing we can do about it.
But users can trust the generated binaries: because we use official codes to generate them, they are safe from the perspective of sources.

JackBoosY commented:
For new ports: 
1. If some architectures are not officially supported, we should add the `supports` field in `vcpkg.json` to filter these architectures. Please use a blacklist instead of a whitelist to avoid unofficial triplets unable to build. See [documentation](https://github.com/microsoft/vcpkg/blob/master/docs/specifications/manifests.md#definitions).
2. If support cannot be added in some official triplets for the time being, we should set these triplets to `fail` in _scripts/ci.baseline.txt_. You can check [this file](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) to get some examples.

plq commented:
- Bumped nss to 3.73.1 where an important security update was released.
- Forked and patched GYP so now it supports msvc 17

JackBoosY commented:
Please get failure logs [here](https://dev.azure.com/vcpkg/public/_build/results?buildId=65290&view=artifacts&pathAsName=false&type=publishedArtifacts).

plq commented:
- Public headers now go to include/nss (was going to include/nss/nss)
- Private headers now go to include/nss/private (was going to include/nss/private/nss)
- Tools (.exe files) are now being copied to their designated locations
- vcpkg.json file was updated for NSS
- scripts/ci.baseline.txt was modified to ignore non-tested triplets
- CI fails because ninja can't be found. I have it installed here: https://github.com/microsoft/vcpkg/pull/21281/commits/c3d41e15a99aafeb5b30beae40626c7aa8186fa1#diff-960b668db44aad86a4013785856a99f980c4ee1337669d221b12aa17ca7cc130R23

    I don't understand the problem here. Any input @JackBoosY ?

Thanks!

plq commented:
> Please get failure logs [here](https://dev.azure.com/vcpkg/public/_build/results?buildId=65290&view=artifacts&pathAsName=false&type=publishedArtifacts).

Yes, I looked at them. The error is very simple:

```
Building NSS requires an installation of ninja: https://ninja-build.org/
```

How do I make sure ninja is found? This build without any problems on my system.

I'm using https://developer.microsoft.com/en-us/windows/downloads/virtual-machines/

plq commented:
https://groups.google.com/a/mozilla.org/g/dev-tech-crypto/c/aV_2AJaPimI

JackBoosY commented:
I can't repro this on my local machine which using Visual Studio 2017.

plq commented:
> I can't repro this on my local machine which using Visual Studio 2017.

Can you be a bit more specific? Is it working or is it not working?

JackBoosY commented:
@plq It built fine on my local machine. I will try to improve the code.

Commit:
Improve code

Commit:
revert baseline change

Commit:
Add supports field

Commit:
improve code

Commit:
improve code

Commit:
version

JackBoosY commented:
We can convert mozbuild part to `vcpkg_*_mozbuild`, but only these two ports are using them, which is not worth it.

plq commented:
Vanilla GYP doesn't work with MSVC 17 aka Visual Studio 2022. You need to use my fork (https://github.com/plq/gyp/) for it to work.

Commit:
Update doc

Commit:
Update doc

plq commented:
GYP is abandoned, there is no upstream to approve my (or anybody's) changes.

JackBoosY commented:
I checked the source, and I found configure/make in it. Maybe we can swith to `vcpkg_*_make`.

plq committed:
[nss] Switch to gyp-nss fork to fix msvc 17 build

plq committed:
[nss] Switch to gyp-nss fork to fix msvc 17 build

plq commented:
Readme says the make-based build system is legacy and can be removed in future, that's why I went with GYP-based one.

I don't think GYP belongs in vcpkg_fetch_acquire because:

1. It's abandoned, nobody is using it
2. Vanilla version doesn't work with nss/windows/msvc17 combination

Frankly, I don't see the difference between 

    https://github.com/plq/gyp/archive/b3177c3f6c2c45a8ca098ae0f0ebb4536c624762.zip

and

    https://github.com/chromium/gyp/archive/d6c5dd51dc3a60bf4ff32a5256713690a1a10376.zip

So you need to either bend your official tools policy for this one or drop this pull request.

JackBoosY commented:
@plq If the upstream approve your changes, I think we can use your repo 😃

plq commented:
![image](https://user-images.githubusercontent.com/79872/148164311-62267b97-b68a-4648-8f98-1731043532de.png)

plq commented:
hahaha seriously though, GYP is abandoned and there is no upstream. Chrome switched to GN, NodeJS forked it as gyp-next and maybe I should name this as gyp-nss and be done with it?

JackBoosY commented:
@plq Also you should use the same license with gyp.

plq committed:
[nss] Improve code

plq committed:
[nss] Improve code

plq commented:
> @plq Also you should use the same license with gyp.

No problem. https://github.com/plq/gyp-nss

plq commented:
So, are we done? Should I flag this as "ready for review"?

JackBoosY commented on scripts/cmake/vcpkg_find_acquire_program.cmake:
```diff
@@ -566,6 +568,23 @@ function(vcpkg_find_acquire_program program)
+        set(paths_to_search "${DOWNLOADS}/tools/gyp-nss/gyp-nss-${program_version}")
+        set(download_urls "https://github.com/plq/gyp-nss/archive/${program_version}.zip")
+        set(download_filename "gyp-nss-${program_version}.zip")
+        set(download_sha512 8996d71ae5e9c3dc365df931dd5363b4e57b9924cffc39d0ab0e696633f8c87398025d396fb1a97feba6d96eeb2ebc233d6cc9d1e77d7d9651806473818faf7f)
```
```suggestion
        set(download_sha512 7cd05e1bdcdb579e8226ecae2e925285e164349927f60350b87703afe9cbdc308f044bc9f6455318f99778b7b49304003aab47f6c587a13e6fbdaaa1533c558d )
```

dg0yt commented on scripts/cmake/vcpkg_find_acquire_program.cmake:
```diff
@@ -566,6 +568,23 @@ function(vcpkg_find_acquire_program program)
+        set(tool_subdirectory "${program_version}")
+        set(search_names "mozmake")
+        set(paths_to_search "${DOWNLOADS}/tools/mozbuild/${program_version}/bin")
+        set(download_urls "https://ftp.mozilla.org/pub/mozilla/libraries/win32/MozillaBuildSetup-${program_version}.exe")
```
So this downloads Windows mozbuild on Linux and macOS?

Changing this scripts triggers rebuilding many vcpkg ports in CI.

plq commented:
> _This is a new experimental fast check for PR issues. Please let us know if this bot is helpful!_
> PRs must add only one version and must not modify any published versions

for the record, this is very annoying

JackBoosY commented:
> > _This is a new experimental fast check for PR issues. Please let us know if this bot is helpful!_
> > PRs must add only one version and must not modify any published versions
> 
> for the record, this is very annoying

That's for the version control.

plq committed:
[nss] Bump to 3.74

plq committed:
[nss] Bump version to 3.74

JackBoosY commented:
Since the system does not necessarily include lapack, I hope to still use lapack in vcpkg.

JackBoosY commented:
When building nss:
```
[749/842] ninja -t msvc -e environment.x64 -- "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe" /nologo /showIncludes /FC @obj\nss-tool\hw-support.hw-support.obj.rsp /c ..\..\nss-tool\hw-support.c /Foobj\nss-tool\hw-support.hw-support.obj /Fdobj\nss-tool\hw-support.c.pdb 
FAILED: obj/nss-tool/hw-support.hw-support.obj 
ninja -t msvc -e environment.x64 -- "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe" /nologo /showIncludes /FC @obj\nss-tool\hw-support.hw-support.obj.rsp /c ..\..\nss-tool\hw-support.c /Foobj\nss-tool\hw-support.hw-support.obj /Fdobj\nss-tool\hw-support.c.pdb 
d:\buildtrees\nss\x64-windows-dbg\nss\lib\freebl\blapii.h(83): error C2143: syntax error: missing ')' before '*'
d:\buildtrees\nss\x64-windows-dbg\nss\lib\freebl\blapii.h(83): error C2143: syntax error: missing '{' before '*'
d:\buildtrees\nss\x64-windows-dbg\nss\lib\freebl\blapii.h(83): error C2059: syntax error: 'type'
d:\buildtrees\nss\x64-windows-dbg\nss\lib\freebl\blapii.h(83): error C2059: syntax error: ')'
```
And please apply [this patch](https://dev.azure.com/vcpkg/_apis/resources/Containers/15463507/format.diff?itemPath=format.diff%2Fformat.diff).

plq commented:
> When building nss:
> 
> ```
> FAILED: obj/nss-tool/hw-support.hw-support.obj 

Interesting, builds fine here

> And please apply [this patch](https://dev.azure.com/vcpkg/_apis/resources/Containers/15463507/format.diff?itemPath=format.diff%2Fformat.diff).

```
{"$id":"1","innerException":null,"message":"TF400813: The user 'Windows Live ID\\<redacted>' is not authorized to access this resource.","typeName":"Microsoft.TeamFoundation.Framework.Server.UnauthorizedRequestException, Microsoft.TeamFoundation.Framework.Server","typeKey":"UnauthorizedRequestException","errorCode":0,"eventId":3000}
```

JackBoosY commented:
[unzip this file](https://github.com/microsoft/vcpkg/files/7827449/format.20.zip) to get patch.

plq commented:
> [unzip this file](https://github.com/microsoft/vcpkg/files/7827449/format.20.zip) to get patch.

done

plq commented:
> When building nss:
> 
> ```
> [749/842] ninja -t msvc -e environment.x64 -- "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe" /nologo /showIncludes /FC @obj\nss-tool\hw-support.hw-support.obj.rsp /c ..\..\nss-tool\hw-support.c /Foobj\nss-tool\hw-support.hw-support.obj /Fdobj\nss-tool\hw-support.c.pdb 
> FAILED: obj/nss-tool/hw-support.hw-support.obj 
> ninja -t msvc -e environment.x64 -- "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe" /nologo /showIncludes /FC @obj\nss-tool\hw-support.hw-support.obj.rsp /c ..\..\nss-tool\hw-support.c /Foobj\nss-tool\hw-support.hw-support.obj /Fdobj\nss-tool\hw-support.c.pdb 
> d:\buildtrees\nss\x64-windows-dbg\nss\lib\freebl\blapii.h(83): error C2143: syntax error: missing ')' before '*'
> d:\buildtrees\nss\x64-windows-dbg\nss\lib\freebl\blapii.h(83): error C2143: syntax error: missing '{' before '*'
> d:\buildtrees\nss\x64-windows-dbg\nss\lib\freebl\blapii.h(83): error C2059: syntax error: 'type'
> d:\buildtrees\nss\x64-windows-dbg\nss\lib\freebl\blapii.h(83): error C2059: syntax error: ')'
> ```
> 
> And please apply [this patch](https://dev.azure.com/vcpkg/_apis/resources/Containers/15463507/format.diff?itemPath=format.diff%2Fformat.diff).

I dug this a bit. Here's my theory:

- Here's blapii.h: https://github.com/nss-dev/nss/blob/ae9fc118f98573a2e04dc736e59326b8d09bc60c/lib/freebl/blapii.h
- Despite the appearances, MSVC complains about mp_int being undefined.
- mp_int is supposed to be defined in mpi.h
- Here's mpi.h: https://github.com/nss-dev/nss/blob/ae9fc118f98573a2e04dc736e59326b8d09bc60c/lib/freebl/mpi/mpi.h
- MPI stands for two very separate things: Multi-Precision Integer [Library] and Message Passing Interface.
- If you have one of the Message Passing Interface implementations installed, the ``include "mpi.h"`` line finds it, because (I suspect) it's installed in includedir root, which makes it higher priority than the mpi.h from the nss repository.
- Hence, MSVC reporting syntax error, erm, undefined type error.

JackBoosY commented:
Sorry for late, I will handle this tomorrow. I'm really busy.

JackBoosY commented:
Reproducing now.

JackBoosY commented:
Can't reproduce it locally, curious why it happened.

JackBoosY commented:
```
[730/826] ninja -t msvc -e environment.x86 -- "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x86\cl.exe" /nologo /showIncludes /FC @obj\nss-tool\hw-support.hw-support.obj.rsp /c ..\..\nss-tool\hw-support.c /Foobj\nss-tool\hw-support.hw-support.obj /Fdobj\nss-tool\hw-support.c.pdb 
FAILED: obj/nss-tool/hw-support.hw-support.obj 
ninja -t msvc -e environment.x86 -- "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x86\cl.exe" /nologo /showIncludes /FC @obj\nss-tool\hw-support.hw-support.obj.rsp /c ..\..\nss-tool\hw-support.c /Foobj\nss-tool\hw-support.hw-support.obj /Fdobj\nss-tool\hw-support.c.pdb 
d:\buildtrees\nss\x86-windows-dbg\nss\lib\freebl\blapii.h(83): error C2143: syntax error: missing ')' before '*'
d:\buildtrees\nss\x86-windows-dbg\nss\lib\freebl\blapii.h(83): error C2143: syntax error: missing '{' before '*'
d:\buildtrees\nss\x86-windows-dbg\nss\lib\freebl\blapii.h(83): error C2059: syntax error: 'type'
d:\buildtrees\nss\x86-windows-dbg\nss\lib\freebl\blapii.h(83): error C2059: syntax error: ')'
```
Maybe need to repro it with Visual Studio 2019, will retry now.

plq commented:
to confirm my theory, try to compile that unit with /showIncludes to see whether the correct mpi.h gets included

JackBoosY commented:
> to confirm (or deny) my theory, try to compile that unit with /showIncludes to see whether the correct mpi.h gets included

I don't think so, if the header `mpi.h` was not included, it will report a different compiler error.

plq commented:
> > to confirm (or deny) my theory, try to compile that unit with /showIncludes to see whether the correct mpi.h gets included
> 
> I don't think so, if the header `mpi.h` was not included, it will report a different compiler error.

I'm not saying that mpi.h is not being found, I'm saying that maybe **the wrong mpi.h is being found**

JackBoosY commented:
All the `mpi.h`:
- blaze:x86-windows:blaze/config/MPI.h
- blaze:x86-windows:blaze/system/MPI.h
- msmpi:x86-windows:mpi.h
- mathgl:x86-windows:mgl2/mpi.h

Which header should we use?

plq commented:
None, nss comes with its own mpi.h. see my comment above: https://github.com/microsoft/vcpkg/pull/21281#issuecomment-1007448956

plq commented:
> msmpi:x86-windows:mpi.h

This can be the only culprit as it installs itself in the include dir root.

plq commented:
I can confirm that the following breaks the nss build:

```
cd path/to/vcpkg
git clean -fdx 
.\bootstrap-vcpkg.bat
.\vcpkg.exe install msmpi (fails because need to install additional package mentioned in the console output)
.\vcpkg.exe install msmpi (succeeds this time)
.\vcpkg.exe install nss (fails with the above error)
```
```

JackBoosY commented:
> msmpi

I see. Since the source include `mpi.h` directly, it finds msmpi's header first.

JackBoosY commented:
We should add relative paths to all the places that directly include `mpi.h`, and add the corresponding absolute paths to `include_dirs` in the configuration file (nss_tool.gyp).

Can you do that?

plq commented:
Sure, let me see what I can do

plq committed:
[nss] Bump to 3.75

plq committed:
[nss] Fix mpi.h confusion

When GYP generates rules for ninja, it puts nspr include dirs before nss's
own include dirs. When the msmpi package is installed, this causes
mpi.h from msmpi to be found, instead of multi-precision integer library
that comes bundled with nss.

This patch uses nspr headers from include/nspr subdirectory which sidesteps
the mpi.h from the include root

JackBoosY commented:
Waiting for pipeline test results.

plq commented:
wow, it's all green. didn't expect that tbh

JackBoosY commented on scripts/ci.baseline.txt:
```diff
@@ -834,6 +834,17 @@ nng:arm-uwp=fail
 nng:x64-uwp=fail
 nrf-ble-driver:arm-uwp=fail
 nrf-ble-driver:x64-uwp=fail
+nspr:arm64-windows=fail
+nspr:arm-uwp=fail
+nspr:x64-osx=fail
+nspr:x64-uwp=fail
+nspr:x64-windows-static=fail
+nss:arm64-windows=fail
+nss:arm-uwp=fail
+nss:x64-osx=fail
+nss:x64-uwp=fail
+nss:x64-linux=fail
+nss:x64-windows-static=fail
```
```suggestion
```
We don't need this since you have already added `supports`.

plq replied:
done

JackBoosY commented on ports/nspr/portfile.cmake:
```diff
@@ -0,0 +1,97 @@
+)
+
+vcpkg_find_acquire_program(MOZBUILD)
+get_filename_component(MOZBUILD_ROOT ${MOZBUILD} DIRECTORY)
```
```suggestion
get_filename_component(MOZBUILD_ROOT "${MOZBUILD}" DIRECTORY)

```

JackBoosY commented on ports/nspr/portfile.cmake:
```diff
@@ -0,0 +1,97 @@
+get_filename_component(MOZBUILD_ROOT "${MOZBUILD_ROOT}" PATH)
+
+set(MOZBUILD_BINDIR "${MOZBUILD_ROOT}/bin")
+vcpkg_add_to_path(${MOZBUILD_BINDIR})
```
```suggestion
vcpkg_add_to_path("${MOZBUILD_BINDIR}")

```

JackBoosY commented on ports/nspr/portfile.cmake:
```diff
@@ -0,0 +1,97 @@
+    "--disable-debug-rtl"
+)
+
+if (VCPKG_TARGET_IS_WINDOWS)
+    set(ENV{CC}  "cl")
+    set(ENV{CXX} "cl")
+    set(ENV{LD}  "link")
+endif()
```
We may not need to do that since `vcpkg_configure_make` will handle them automaticly.

plq replied:
I removed these and it didn't seem to break anything.

JackBoosY commented on ports/nspr/portfile.cmake:
```diff
@@ -0,0 +1,97 @@
+        file(RENAME "${CURRENT_PACKAGES_DIR}/lib/${name}" "${CURRENT_PACKAGES_DIR}/bin/${name}")
+    endforeach()
+
+    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/share")
```
Which files are installed in _share_ folder?

plq replied:
This file: https://github.com/nss-dev/nss/blob/master/COPYING named "copyright"

plus there is a file named "vcpkg_abi_info"

plq replied:
What should we do with these?

JackBoosY replied:
We should keep this folder and change the copyright installtion to:
```cmake
file(RENAME "${CURRENT_PACKAGES_DIR}/share/COPYING" "${CURRENT_PACKAGES_DIR}/share/copyright")
```

JackBoosY commented on ports/nss/portfile.cmake:
```diff
@@ -0,0 +1,225 @@
+# setup mozbuild for windows
+if (VCPKG_TARGET_IS_WINDOWS)
+    vcpkg_find_acquire_program(MOZBUILD)
+    get_filename_component(MOZBUILD_ROOT ${MOZBUILD} DIRECTORY)
```
```suggestion
    get_filename_component(MOZBUILD_ROOT "${MOZBUILD}" DIRECTORY)

```

JackBoosY commented on ports/nss/portfile.cmake:
```diff
@@ -0,0 +1,225 @@
+endif()
+
+vcpkg_find_acquire_program(GYP_NSS)
+get_filename_component(GYP_NSS_ROOT ${GYP_NSS} DIRECTORY)
```
```suggestion
get_filename_component(GYP_NSS_ROOT "${GYP_NSS}" DIRECTORY)
```

JackBoosY commented on ports/nss/portfile.cmake:
```diff
@@ -0,0 +1,225 @@
+message(STATUS "Found gyp: ${GYP_NSS}")
+
+# setup paths
+execute_process(
+    COMMAND ${MOZBUILD_BASH} -c pwd
+    WORKING_DIRECTORY ${CURRENT_INSTALLED_DIR}/include
+    OUTPUT_VARIABLE VCPKG_INCLUDEDIR
+    OUTPUT_STRIP_TRAILING_WHITESPACE
+)
+message(STATUS "Using headers from: ${VCPKG_INCLUDEDIR} arch: ${VCPKG_TARGET_ARCHITECTURE}")
+
+execute_process(
+    COMMAND ${MOZBUILD_BASH} -c pwd
+    WORKING_DIRECTORY ${CURRENT_INSTALLED_DIR}/lib
+    OUTPUT_VARIABLE VCPKG_LIBDIR
+    OUTPUT_STRIP_TRAILING_WHITESPACE
+)
+message(STATUS "Using libraries from: ${VCPKG_LIBDIR} arch: ${VCPKG_TARGET_ARCHITECTURE}")
```
Can we use `${CURRENT_BUILDTREES_DIR}` as the base directory and add these paths?

plq replied:
The last two lines of the below quoted build log are the outputs of the message() calls from the mentioned block of code:

```
-- Found gyp: L:/src/github/plq/vcpkg/downloads/tools/gyp-nss/gyp-nss-b3177c3f6c2c45a8ca098ae0f0ebb4536c624762/gyp
-- Using headers from: /l/src/github/plq/vcpkg/installed/x86-windows/include arch: x86
-- Using libraries from: /l/src/github/plq/vcpkg/installed/x86-windows/lib arch: x86
```

As you see, ``bash -c pwd`` returns unix-style paths. AFAICT that's what the nss build system expects. So I think we should keep this as is.

JackBoosY commented on ports/nss/portfile.cmake:
```diff
@@ -0,0 +1,225 @@
+
+# Debug tools
+
+vcpkg_copy_tools(
+    TOOL_NAMES
+        "certutil"
+        "cmsutil"
+        "crlutil"
+        "hw-support"
+        "modutil"
+        "nss"
+        "pk12util"
+        "pwdecrypt"
+        "shlibsign"
+        "signtool"
+        "signver"
+        "ssltap"
+        "symkeyutil"
+        "validation"
+    SEARCH_DIR "${VCPKG_BINARY_DIR}-dbg/dist/Debug/bin/"
+    DESTINATION "${CURRENT_PACKAGES_DIR}/debug/tools/${PORT}"
+)
```
We don't need the debug version of these tools,

plq replied:
OK, I removed them

plq committed:
[nss] add quotes around paths

Co-authored-by: Jack·Boos·Yu <47264268+JackBoosY@users.noreply.github.com>

plq committed:
[nss] remove debug versions of tools

plq committed:
[nspr] Remove redundant environment variable assignments

plq commented:
Hey @JackBoosY, thanks for the review. I think I addressed all of your points. So back to you :) 🎊🎈 🎊

plq commented:
There were some weird instructions in the nspr ebuild. I think this final version addresses all of your concerns

JackBoosY commented on scripts/cmake/vcpkg_find_acquire_program.cmake:
```diff
@@ -566,6 +568,23 @@ function(vcpkg_find_acquire_program program)
             set(apt_package_name pkg-config)
             set(paths_to_search "/bin" "/usr/bin" "/usr/local/bin")
         endif()
+    elseif (program STREQUAL "MOZBUILD")
+        set(program_name mozbuild)
+        set(program_version 3.3)
+        set(tool_subdirectory "${program_version}")
+        set(search_names "mozmake")
+        set(paths_to_search "${DOWNLOADS}/tools/mozbuild/${program_version}/bin")
+        set(download_urls "https://ftp.mozilla.org/pub/mozilla/libraries/win32/MozillaBuildSetup-${program_version}.exe")
+        set(download_filename "mozbuild-${program_version}.7z.exe")
+        set(download_sha512 ac33d15dd9c974ef8ad581f9b414520a9d5e3b9816ab2bbf3e305d0a33356cc22c356cd9761e64a19588d17b6c13f124e837cfb462a36b8da898899e7db22ded)
+    elseif (program STREQUAL "GYP_NSS")
+        set(program_name "gyp-nss")
+        set(program_version b3177c3f6c2c45a8ca098ae0f0ebb4536c624762)
+        set(search_names "gyp")
+        set(paths_to_search "${DOWNLOADS}/tools/gyp-nss/gyp-nss-${program_version}")
+        set(download_urls "https://github.com/plq/gyp-nss/archive/${program_version}.zip")
+        set(download_filename "gyp-nss-${program_version}.zip")
+        set(download_sha512 7cd05e1bdcdb579e8226ecae2e925285e164349927f60350b87703afe9cbdc308f044bc9f6455318f99778b7b49304003aab47f6c587a13e6fbdaaa1533c558d )
```
Yes, that's the modern way, but I think we're okay with that now.

plq replied:
That's @JackBoosY 's code, not mine, despite the appearances.

dg0yt replied:
I mentioned this issue very early. I'm not really okay with it. Changes to this script are expensive, and each addition makes it more expensive. And it is expensive for all platforms, even if the current changes are only for a single platform.

*Please stop addition of tools here.*

JackBoosY replied:
Fine, I will create a new tool port instead of this.

JackBoosY commented:
Close and reopen this PR to re-active the CLA check.

Commit:
Merge branch 'public' of https://github.com/plq/vcpkg into public

Commit:
Merge branch 'master' of https://github.com/microsoft/vcpkg into public

Commit:
Add dependencies vcpkg-tool-mozbuild and vcpkg-tool-gyp-nss

Commit:
clean doc

JackBoosY commented on ports/vcpkg-tool-gyp-nss/portfile.cmake:
```diff
@@ -0,0 +1,25 @@
+set(PROG_VERSION b3177c3f6c2c45a8ca098ae0f0ebb4536c624762)
+set(ARCHIVE gyp-nss-${PROG_VERSION})
+
+if(VCPKG_CROSSCOMPILING)
+    message(FATAL_ERROR "This is a host only port!")
+endif()
```
I assume this is not needed, right?

Commit:
version

JackBoosY commented:
Looks passed.
@Neumann-A @dg0yt Please review again.

JackBoosY commented:
Will add `license` after review.

plq commented:
NSS 3.76 was released: https://groups.google.com/a/mozilla.org/g/dev-tech-crypto/c/W13LB93wep4/m/v3yN3ISSBAAJ

Should I wait until this gets merged and do it in a separate PR or just do it?

JackBoosY commented:
@plq Let's wait for merge this first.

plq committed:
Merge branch 'master' into public

BillyONeal commented on ports/nspr/portfile.cmake:
```diff
@@ -0,0 +1,88 @@
+    list(APPEND OPTIONS "--enable-64bit")
+elseif(VCPKG_TARGET_ARCHITECTURE STREQUAL "x86")
+    list(APPEND OPTIONS "--disable-64bit")
+endif()
```
```suggestion
else()
    message(FATAL_ERROR "Unrecognized VCPKG_TARGET_ARCHITECTURE.")
endif()

```

I see you don't care about arm et-al in your comments but it's still good to put defenses in here in case someone in the future tries changing the supports and doesn't realize they need to resolve that problem.

BillyONeal commented on ports/nspr/portfile.cmake:
```diff
@@ -0,0 +1,88 @@
+endif()
+
+# Debug
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/share")
+file(GLOB BIN_DEBUG "${CURRENT_PACKAGES_DIR}/debug/lib/*.dll" "${CURRENT_PACKAGES_DIR}/debug/lib/*.pdb")
+list(LENGTH BIN_DEBUG BIN_DEBUG_SIZE)
+if (BIN_DEBUG_SIZE GREATER 0)
+    file(MAKE_DIRECTORY "${CURRENT_PACKAGES_DIR}/debug/bin")
+
+    foreach(path IN LISTS BIN_DEBUG)
+        get_filename_component(name "${path}" NAME)
+        file(RENAME "${CURRENT_PACKAGES_DIR}/debug/lib/${name}" "${CURRENT_PACKAGES_DIR}/debug/bin/${name}")
+    endforeach()
+endif()
```
This block should be guarded by:

```
if (NOT VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "debug")
```

BillyONeal replied:
Also shouldn't we build debug first for consistency with other things or is there a reason to build release first?

BillyONeal commented on ports/nss/portfile.cmake:
```diff
@@ -0,0 +1,203 @@
+elseif(VCPKG_TARGET_ARCHITECTURE STREQUAL "x86")
+    list(APPEND OPTIONS "--target=ia32")
+else()
+    message(FATAL "Unsupported arch: %{VCPKG_TARGET_ARCHITECTURE}")
```
```suggestion
    message(FATAL_ERROR "Unsupported arch: ${VCPKG_TARGET_ARCHITECTURE}")

```

BillyONeal commented on ports/vcpkg-tool-mozbuild/vcpkg.json:
```diff
@@ -0,0 +1,7 @@
+  "version": "3.3",
+  "description": "mozbuild is a Python package providing functionality used by Mozilla’s build system",
+  "homepage": "https://firefox-source-docs.mozilla.org/build/buildsystem/mozbuild/index.html",
+  "supports": "native & !x86 & windows"
```
```suggestion
  "supports": "native & x64 & windows"
```

BillyONeal commented on ports/nss/portfile.cmake:
```diff
@@ -0,0 +1,203 @@
+set(VCPKG_BINARY_DIR "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}")
+
+# build debug
+message(STATUS "Copying sources to debug build dir ...")
+file(COPY "${SOURCE_PATH}/nss" DESTINATION "${VCPKG_BINARY_DIR}-dbg")
+message(STATUS "Building debug ...")
+vcpkg_execute_required_process(
+    COMMAND ${MOZBUILD_BASH} ./build.sh ${OPTIONS}
+    WORKING_DIRECTORY ${VCPKG_BINARY_DIR}-dbg/nss
+    LOGNAME build-${TARGET_TRIPLET}${short_buildtype}
+)
```
Ditto please guard with

```
if (NOT VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "debug")
```

BillyONeal commented:
I'm pretty sure this is pretty much good, thank you so much for your contribution despite it taking a long time to get 'through the tubes'! I made only nitpick comments and 1 question.

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into public

BillyONeal committed:
Remove unnecessary vcpkg_crosscompiling block on vcpkg-tool-gyp-nss.

Guard for VCPKG_BUILD_TYPE
Improve error reporting for unsupported VCPKG_TARGET_ARCHITECTURE
Declare support for x64 rather than !x86 (given that the URI in question only has x64 binaries)
Fix version numbers to use 'version'.

BillyONeal commented:
I just made the nitpick changes. Thanks for your contribution!

JackBoosY commented:
Looks all good, thanks for @BillyONeal review.

BillyONeal commented:
Thanks for your contribution!

plq commented:
If it weren't for the fabulous test bench you guys built, this would have taken even longer. Thanks a lot for the help, kudos :)
